### PR TITLE
Implement the comma command to manage user-defined tables ##util

### PR DESCRIPTION
**Pull request by trufae**
Continuation of: https://github.com/radareorg/radare2/pull/16824

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: libr/core/cmd.c [hunk 1]
```diff
@@ -1266,6 +1282,241 @@ static int cmd_stdin(void *data, const char *input) {
...
+			r_list_append (t->rows, row);
+		}
+	}
+	r_list_free (lines);
+	free (data);
+}
+
+static int cmd_table(void *data, const char *input) {
+	RCore *core = (RCore*)data;
+	static RTable *t = NULL;
```
Comment (trufae) [inline libr/core/cmd.c:1387]: NOTE TO SELF: move into core->table

File: libr/core/cmd.c [hunk 2]
```diff
@@ -1266,6 +1282,241 @@ static int cmd_stdin(void *data, const char *input) {
...
+				r_cons_printf ("%s\n", ts);
+				free (ts);
+			}
+		}
+		break;
+	case 0:
+		// print table
+		{
+			char *ts = r_table_tofancystring (t);
+			// char *ts = r_table_tostring (t);
```
Comment (trufae) [inline libr/core/cmd.c:1487]: better to just call tostring()

File: libr/util/table.c [hunk 3]
```diff
@@ -438,6 +438,37 @@ R_API char *r_table_tostring(RTable *t) {
...
+	}
+	r_list_foreach (t->cols, iter, col) {
+		r_strbuf_appendf (sb, " %s",  col->name);
+	}
+	r_strbuf_append (sb, "\n");
+
+	r_list_foreach (t->rows, iter, row) {
+		char *item;
+		int c = 0;
+	r_strbuf_appendf (sb, ",r ");
```
Comment (trufae) [inline libr/util/table.c:460]: wrong indent

**Review by trufae — COMMENTED** [review]
.

**Comment by XVilka** [comment]
I propose the following syntax:
`[` as a command instead of `,`, and put the rest of the command inside `[]`:
```
[0x00000000]> [h xxd foo bar cow]
.--------------------.
| foo  | bar  | cow  |
)--------------------(
`--------------------'
[0x00000000]> [r 123 456 789]
.--------------------.
| foo  | bar  | cow  |
)--------------------(
| 123  | 456  | 789  |
`--------------------'
```

**Comment by trufae** [comment]
i like the [] idea, but imnot sure if that will work the same for other commands, and this will be probably more complicated to support In TS. imho im good to merge this when CI decides to go green, and i can switch to [] to all the other commands using rtable instead of , or t. wdyt?

**Comment by XVilka** [comment]
@ret2libc what do you think? I don't think using `[...]` syntax will break the tree-sitter parser.

**Comment by XVilka** [comment]
Also it's red:
```
[XX] db/cmd/cmd_table comma table
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'afr
e scr.utf8=true
aflt > .tt
cat .tt~?
, .tt
rm .tt
,~?
' bins/mach0/mac-ls
-- stdout
@@ -1,2 +1,5 @@
 37
-37
+
+
+
+0

[XX] db/cmd/cmd_table comma table
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'afr
e scr.utf8=false
aflt > .tt2
cat .tt2~?
, .tt2
rm .tt2
,~?
' bins/mach0/mac-ls
-- stdout
@@ -1,2 +1,5 @@
 37
-37
+
+
+
+0
```

**Comment by ret2libc** [comment]
@trufae @XVilka what is exactly the syntax you want to use with `[..]`? Also, I don't think it makes sense to merge `,` and change everything few commits later with `[`. Let's just choose now what we are going to use and let's go with it from the beginning.

**Comment by XVilka** [comment]
@ret2libc see https://github.com/radareorg/radare2/pull/17702#issuecomment-698254970

File: libr/core/cmd.c [hunk 4]
```diff
@@ -7086,6 +7332,7 @@ R_API void r_core_cmd_init(RCore *core) {
 		{"*",        "pointer read/write", cmd_pointer, NULL, &pointer_help},
 		{"-",        "open cfg.editor and run script", cmd_stdin, NULL, &stdin_help},
 		{".",        "interpret", cmd_interpret, NULL, &interpret_help},
+		{",",        "create and manipulate tables", cmd_table, NULL, &table_help},
```
Comment (ret2libc) [inline libr/core/cmd.c:7335]: this probably cannot work as `table_help` is not of type `RCmdDescHelp`.

File: libr/core/cmd.c [hunk 5]
```diff
@@ -153,6 +153,22 @@ static const char *help_msg_star[] = {
 	NULL
 };
 
+static const char *table_help[] = {
+	"Usage:", ",[/] [file]", "# load table data",
+	",", "", "display table",
+	".", "$foo", "aflt > $foo (files starting with '$' are saved in memory)",
```
Comment (ret2libc) [inline libr/core/cmd.c:159]: why the `.` commands? I don't think they belong here.

File: libr/core/cmd.c [hunk 6]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+	}
+	r_list_free (lines);
+	free (data);
+}
+
+static int cmd_table(void *data, const char *input) {
+	RCore *core = (RCore*)data;
+
+	if (!core->table) {
+		core->table = r_table_new ();
```
Comment (ret2libc) [inline libr/core/cmd.c:1389]: could you just initialize it together with all `RCore` fields and avoid the lazy initialization?

File: libr/core/cmd.c [hunk 7]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+	}
+
+	switch (*input) {
```
Comment (ret2libc) [inline libr/core/cmd.c:1392]: could you split each case like how it is done in https://github.com/radareorg/radare2/blob/master/libr/core/cmd_write.c#L1981 ?

File: libr/core/cmd.c [hunk 8]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+				}
+				const char *type_name = (type_char == 's')
+					? "string": "number";
+				RTableColumnType *typeString = r_table_type (type_name);
+				r_table_add_column (core->table, typeString, s, 0);
+				i++;
+			}
+			r_list_free (list);
+			free (format);
+	//		free (s);
```
Comment (ret2libc) [inline libr/core/cmd.c:1416]:
```suggestion
			free (s);
```

File: libr/core/cmd.c [hunk 9]
```diff
@@ -153,6 +153,22 @@ static const char *help_msg_star[] = {
...
+	",", "", "display table",
+	".", "$foo", "aflt > $foo (files starting with '$' are saved in memory)",
+	". ", " [table-query]", "filter and print table. See ,? for more details",
+	"..", " file.csv", "load the csv file into a table",
+	",-", "", "reset table",
+	",/", "?", "query/filter current table",
+	",*", ">$foo", "print table as r2 commands",
+	",j", "", "print table in json format",
+	",h", "xxd foo bar cow", "define header column names and types",
+	",r", "xxd 1 2 foo", "adds a row using the given format string",
```
Comment (ret2libc) [inline libr/core/cmd.c:167]: `xxd` is probably not required to add a row, just the elements are used AFAIS

File: libr/core/cmd.c [hunk 10]
```diff
@@ -153,6 +153,22 @@ static const char *help_msg_star[] = {
...
+	"Usage:", ",[/] [file]", "# load table data",
+	",", "", "display table",
+	".", "$foo", "aflt > $foo (files starting with '$' are saved in memory)",
+	". ", " [table-query]", "filter and print table. See ,? for more details",
+	"..", " file.csv", "load the csv file into a table",
+	",-", "", "reset table",
+	",/", "?", "query/filter current table",
+	",*", ">$foo", "print table as r2 commands",
+	",j", "", "print table in json format",
+	",h", "xxd foo bar cow", "define header column names and types",
```
Comment (ret2libc) [inline libr/core/cmd.c:166]:
```suggestion
	",h", " xxd foo bar cow", "define header column names and types",
```
same for other lines

File: libr/util/table.c [hunk 11]
```diff
@@ -462,6 +462,37 @@ R_API char *r_table_tosimplestring(RTable *t) {
...
+	}
+	r_list_foreach (t->cols, iter, col) {
+		r_strbuf_appendf (sb, " %s",  col->name);
+	}
+	r_strbuf_append (sb, "\n");
+
+	r_list_foreach (t->rows, iter, row) {
+		char *item;
+		int c = 0;
+		r_strbuf_appendf (sb, ",r ");
```
Comment (ret2libc) [inline libr/util/table.c:484]:
```suggestion
		r_strbuf_appendf (sb, ",r");
```

File: libr/util/table.c [hunk 12]
```diff
@@ -462,6 +462,37 @@ R_API char *r_table_tosimplestring(RTable *t) {
...
+		r_list_foreach (row->items, iter2, item) {
+			RTableColumn *col = r_list_get_n (t->cols, c);
```
Comment (ret2libc) [inline libr/util/table.c:486]: why is this required? I think that if a row has 3 items, there are at least 3 columns in the table, otherwise something would be very wrong in the table. Am I missing anything?

File: libr/core/cmd.c [hunk 13]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+				free (ts);
+			}
+		}
+		break;
+	case 'j':
+		// print table
+		if (input[1] == ' ') {
+			const char *q = r_str_trim_head_ro (input + 2);
+			if (r_table_query (core->table, q)) {
+				char *ts = r_table_tojson (core->table);
```
Comment (ret2libc) [inline libr/core/cmd.c:1495]: this code is the same as few lines below. You can invert the `if` and write this code just one time and get it executed in both cases (input[1] ==  ' ' and the else case).

File: libr/core/cmd.c [hunk 14]
```diff
@@ -153,6 +153,22 @@ static const char *help_msg_star[] = {
 	NULL
 };
 
+static const char *table_help[] = {
+	"Usage:", ",[/] [file]", "# load table data",
+	",", "", "display table",
+	".", "$foo", "aflt > $foo (files starting with '$' are saved in memory)",
+	". ", " [table-query]", "filter and print table. See ,? for more details",
+	"..", " file.csv", "load the csv file into a table",
```
Comment (ret2libc) [inline libr/core/cmd.c:161]: I think this is not implemented.

File: libr/core/cmd.c (hunk 5 above)
Comment (ret2libc) [reply libr/core/cmd.c:159]: Maybe a typo

File: libr/core/cmd.c [hunk 15]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+		if (r_table_query (core->table, input + 1)) {
+			char *ts = r_table_tostring (core->table);
+			r_cons_printf ("%s", ts);
+			free (ts);
+		}
+		break;
+	case '$':
+		load_table (core, core->table, input);
+		break;
+	case ' ':
```
Comment (ret2libc) [inline libr/core/cmd.c:1445]: what is the difference between this case and the `/` case? They look the same to me apart from the `r_str_trim_head_ro`.

File: libr/core/cmd.c [hunk 16]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+	case '/':
+		// query here
+		if (r_table_query (core->table, input + 1)) {
+			char *ts = r_table_tostring (core->table);
+			r_cons_printf ("%s", ts);
+			free (ts);
+		}
+		break;
+	case '$':
+		load_table (core, core->table, input);
```
Comment (ret2libc) [inline libr/core/cmd.c:1443]: get the `data` from the alias before calling `load_table` and call it just with the data, so `load_table` doesn't need to perform the loading from files and it can potentially get data from everywhere, not just aliases and files.

File: libr/core/cmd.c [hunk 17]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+				char *ts = r_table_tostring (core->table);
+				if (ts) {
+					r_cons_printf ("%s\n", ts);
+					free (ts);
+				}
+			}
+		}
+		break;
+	case '.':
+		load_table (core, core->table, r_str_trim_head_ro (input + 1));
```
Comment (ret2libc) [inline libr/core/cmd.c:1458]: same here, get the data from the file before calling `load_table` and just pass the data.

File: libr/core/cmd.c [hunk 18]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+		return;
+	}
+	RListIter *iter;
+	char *line;
+	RList *lines = r_str_split_list (data, "\n", 0);
+	bool expect_header = false;
+	bool expect_rows = false;
+	const char *separator = "|";
+	int ncols = 0;
+	r_list_foreach (lines, iter, line) {
```
Comment (ret2libc) [inline libr/core/cmd.c:1307]: since the table has a well defined format:
```
<line to determine the separator>
<headers>
<line to determine start of rows>
<rows>
```
it's much easier to read if those things are done separately instead of using one big loop with multiple booleans that completely change the behaviour of the loop. In particular, because it seems you can't really go ahead if any of those pieces is missing from the table.

File: libr/core/cmd.c [hunk 19]
```diff
@@ -1266,6 +1282,239 @@ static int cmd_stdin(void *data, const char *input) {
...
+			char *arg;
+			RList *args = r_str_split_list (line + strlen (separator), separator, 0);
+			RListIter *iter2;
+			ncols = 0;
+			if (r_list_length (t->cols) > 0) {
+				eprintf ("Warning: Not re-adding headers. Use ,- to reset the table.\n");
+				continue;
+			}
+			r_list_foreach (args, iter2, arg) {
+				char *s = strchr (arg, ' ');
```
Comment (ret2libc) [inline libr/core/cmd.c:1343]: why is `strchr` needed here?

**Review by ret2libc — CHANGES_REQUESTED** [review]

**Comment by ret2libc** [comment]
> @ret2libc what do you think? I don't think using `[...]` syntax will break the tree-sitter parser.

Mh, I don't like having `]` at the end of the command, it adds no real value and it requires the intervention of the parser for no real reason. If you want to do things like `[...]` there are 2 ways and I don't like either:
1) you change the grammar and parse the table commands at the parser level (so it would be like a special case and not like any other command)
2) you treat it like a regular command, but then `]` would need to be parsed and filtered by the command handler, which is very dirty and bad.

So, all in all, I think I prefer `,` or just `[...`, without the ending `]`.

**Comment by XVilka** [comment]
Ok, then I think just `[...` is good enough.

File: libr/core/cmd.c (hunk 15 above)
Comment (trufae) [reply libr/core/cmd.c:1455]: Yeah, the / command can be removed i think, just using a space is fine

File: libr/core/cmd.c (hunk 13 above)
Comment (trufae) [reply libr/core/cmd.c:1505]: done

File: libr/core/cmd.c (hunk 4 above)
Comment (trufae) [reply libr/core/cmd.c:7372]: oh right, my bad.fixed

File: libr/util/table.c (hunk 12 above)
Comment (trufae) [reply libr/util/table.c:486]: would you prefer a runtime assert?

File: libr/core/cmd.c (hunk 18 above)
Comment (trufae) [reply libr/core/cmd.c:1317]: it skips lines that dont match those rules

File: libr/core/cmd.c (hunk 19 above)
Comment (trufae) [reply libr/core/cmd.c:1353]: to skip the first word

File: libr/core/cmd.c (hunk 8 above)
Comment (trufae) [reply libr/core/cmd.c:1426]: r_str_split_list owns the pointer. this would cause a double free

File: libr/core/cmd.c (hunk 16 above)
Comment (trufae) [reply libr/core/cmd.c:1453]: done

File: libr/core/cmd.c (hunk 15 above)
Comment (trufae) [reply libr/core/cmd.c:1455]: but it allows to show the help.. maybe one of them should be non destructive

Comment (trufae) [reply libr/core/cmd.c:1455]: yeah, done. review again

File: libr/core/cmd.c (hunk 17 above)
Comment (trufae) [reply libr/core/cmd.c:1468]: done

File: libr/core/cmd.c (hunk 7 above)
Comment (trufae) [reply libr/core/cmd.c:1389]: Yes, but, if the reason is readability it will be better to move the whole thing into a separate .c file. if its for api accessibility thats a different topic, all of them or only the ones that hold local vars between braces?

File: libr/core/cmd.c (hunk 14 above)
Comment (trufae) [reply libr/core/cmd.c]: Correct. r_table_from_csv is not implemented and this command just tries to guess the format to load, which from my tests it seems to not work very reliabiliy. ill write tests for this and fix the behaviour

File: libr/core/cmd.c (hunk 9 above)
Comment (trufae) [reply libr/core/cmd.c:167]: thanks

File: libr/core/cmd.c (hunk 6 above)
Comment (trufae) [reply libr/core/cmd.c]: ok

File: libr/core/cmd.c (hunk 7 above)
Comment (trufae) [reply libr/core/cmd.c:1414]: done

**Comment by trufae** [comment]
@XVilka i have been trying to use , and [ and it feels weird that theres no ] at the end. maybe its my OCD. im sticking with comma for now. what do you think?

File: libr/core/cmd.c (hunk 8 above)
Comment (ret2libc) [reply libr/core/cmd.c]: the list generated by `r_str_split_list` doesn't free the string and since you are freeing the list a couple of lines above, if you don't free the string `s` will leak.

File: libr/util/table.c (hunk 12 above)
Comment (ret2libc) [reply libr/util/table.c:496]: if we don't think this is 100% true yeah, maybe better, though I would expect this to be always true, if the table API is ok.

File: libr/core/cmd.c [hunk 20]
```diff
@@ -153,6 +153,22 @@ static const char *help_msg_star[] = {
 	NULL
 };
 
+static const char *cmd_table_help[] = {
+	"Usage:", ",[/] [file]", "# load table data",
+	",", "", "display table",
+	",", "$foo", "aflt > $foo (files starting with '$' are saved in memory)",
+	", ", " [table-query]", "filter and print table. See ,? for more details",
+	",.", " file.csv", "load the csv file into a table",
```
Comment (ret2libc) [inline libr/core/cmd.c:161]: it is not clear that `,.` can be used to load regular files as well, not only .csv

File: libr/core/cmd.c (hunk 18 above)
Comment (ret2libc) [reply libr/core/cmd.c:1304]: This function seems still very messy to me.

File: libr/core/cmd.c (hunk 19 above)
Comment (ret2libc) [reply libr/core/cmd.c:1340]: why do you want to skip it?
```
.-------------------------.
| foo fuffa | bar  | cow  |
)-------------------------(
| ciao      | zoo  | boh  |
`-------------------------'
```
By doing so, if you load the above table `foo fuffa` becomes `fuffa`. Is that done on purpose?

**Review by ret2libc — CHANGES_REQUESTED** [review]
Much better apart from some small things (and the `load_table` function which seems messy to me, but that's just my opinion so don't worry too much about it, just adjust the other stuff).

File: libr/core/cmd.c [hunk 21]
```diff
@@ -1276,6 +1292,208 @@ static int cmd_stdin(void *data, const char *input) {
...
+		display_table (r_table_tocsv (core->table));
+		break;
+	case '*':
+		display_table (r_table_tor2cmds (core->table));
+		break;
+	case 0:
+		display_table (r_table_tofancystring (core->table));
+		break;
+	case 'j':
+		// print table
```
[the remark this answers is not in the document]
Comment (trufae) [reply libr/core/cmd.c:1477]: Fixed

File: libr/core/cmd.c (hunk 7 above)
Comment (trufae) [reply libr/core/cmd.c:1429]: so do you agree on adding cmd_table.c and adding all those static functions in there? @ret2libc

File: libr/util/table.c (hunk 12 above)
Comment (trufae) [reply libr/util/table.c:496]: See line :249 .. it is not reliable so better keep this safe if until we get proper unit and r2 tests for this

File: libr/core/cmd.c [hunk 22]
```diff
@@ -153,6 +153,22 @@ static const char *help_msg_star[] = {
 	NULL
 };
 
+static const char *cmd_table_help[] = {
+	"Usage:", ",[/] [file]", "# load table data",
+	",", "", "display table",
+	",", "$foo", "aflt > $foo (files starting with '$' are saved in memory)",
+	", ", " [table-query]", "filter and print table. See ,? for more details",
```
[the remark this answers is not in the document]
Comment (trufae) [reply libr/core/cmd.c:161]: done

File: libr/core/cmd.c (hunk 7 above)
Comment (ret2libc) [reply libr/core/cmd.c:1428]: It's also ok as-is, don't worry.

**Review by XVilka — CHANGES_REQUESTED** [review]
4 broken tests
```
[XX] db/cmd/cmd_table comma table
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'afr
aflt > $a
cat $a~?
, $a
rm $a
,~?
' bins/mach0/mac-ls
-- stdout
@@ -1,2 +1,5 @@
 37
-37
+
+
+
+0

[XX] db/cmd/cmd_table comma table
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'afr
e scr.utf8=false
aflt > .tt2
cat .tt2~?
, .tt2
rm .tt2
,~?
' bins/mach0/mac-ls
-- stdout
select: Bad file descriptor
@@ -1,2 +1,5 @@
 37
-37
+
+
+
+0

[XX] TIMEOUT db/cmd/write Write misaligns jump instruction, control flow is reanalyzed
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -a x86 -b 64 -e anal.detectwrites=true -Qc 'wx e80200000075020000c3
af
pdr
wa jne 9
pdr
wx e802
pdr
' -
-- stdout
@@ -7,24 +7,3 @@
 | ----------- true: 0x00000009
 \ 0x00000009      c3             ret
 
-/ 10: fcn.00000000 (int64_t arg_2h);
-| ; arg int64_t arg_2h @ rbp+0x2
-| 0x00000000      7507           jne 9
-| ----------- true: 0x00000009  false: 0x00000002
-| 0x00000002      0000           add byte [rax], al
-| 0x00000004      007502         add byte [arg_2h], dh
-| ; CALL XREF from fcn.00000000 @ 
-| 0x00000007      0000           add byte [rax], al
-| ----------- true: 0x00000009
-\ 0x00000009      c3             ret
-
-/ 10: fcn.00000000 (int64_t arg_2h);
-| ; arg int64_t arg_2h @ rbp+0x2
-| 0x00000000      e802000000     call 7
-| 0x00000005      7502           jne 9
-| ----------- true: 0x00000009  false: 0x00000007
-| ; CALL XREF from fcn.00000000 @ 
-| 0x00000007      0000           add byte [rax], al
-| ----------- true: 0x00000009
-\ 0x00000009      c3             ret
-


-- exit status: 
[XX] db/cmd/cmd_table comma table
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'afr
e scr.utf8=true
aflt > .tt
cat .tt~?
, .tt
rm .tt
,~?
' bins/mach0/mac-ls
-- stdout
@@ -1,2 +1,5 @@
 37
-37
+
+
+
+0
```
